drivers/spi/spi_wch.c

```diff
@@ -0,0 +1,259 @@
+			break;
+		}
+		div_val *= 2;
+	}
+	return div;
+}
+
+
+static int spi_wch_configure(const struct device *dev,
+			      const struct spi_config *config)
+{
+	const struct spi_wch_config *cfg = dev->config;
+	struct spi_wch_data *data = dev->data;
+	int err;
+	uint32_t clock_rate;
+	clock_control_subsys_t clk_sys;
+	int8_t div;
+
+	if (spi_context_configured(&data->ctx, config)) {
+		return 0;
+	}
+
+	if (config->operation & SPI_HALF_DUPLEX) {
+		LOG_ERR("Half-duplex not supported");
+		return -ENOTSUP;
+	}
+
+	if (SPI_OP_MODE_GET(config->operation) != SPI_OP_MODE_MASTER) {
+		return -ENOTSUP;
+	}
+
+	if ((config->operation & SPI_MODE_LOOP) != 0U) {
+		return -ENOTSUP;
+	}
+
+	if (SPI_WORD_SIZE_GET(config->operation) != 8) {
+		return -ENOTSUP;
+	}
+
+	// reset control register
+	cfg->regs->CTLR1 = 0;
+	/* CS */
+	if (spi_cs_is_gpio(config)) {
+		cfg->regs->CTLR1 |= SPI_NSS_Soft;
+		cfg->regs->CTLR2 &= ~SPI_CTLR2_SSOE;
+	} else {
+		cfg->regs->CTLR1 |= SPI_NSS_Hard;
+		cfg->regs->CTLR2 |= SPI_CTLR2_SSOE;
+	}
+	/* Master */
+	cfg->regs->CTLR1 |= SPI_Mode_Master;
+	/* Full Duplex */
+	cfg->regs->CTLR1 |= SPI_Direction_2Lines_FullDuplex;
+
+	if ((config->operation & SPI_TRANSFER_LSB) != 0U) {
+		cfg->regs->CTLR1 |= SPI_CTLR1_LSBFIRST;
+	}
+
+	if ((config->operation & SPI_MODE_CPOL) != 0U) {
+		cfg->regs->CTLR1 |= SPI_CPOL_High;
+	}
+
+	if ((config->operation & SPI_MODE_CPHA) != 0U) {
+		cfg->regs->CTLR1 |= SPI_CPHA_2Edge;
+	}
+
+	/* 8 bits per transfer */
+	cfg->regs->CTLR1  &= ~(SPI_CTLR1_DFF);
+
+
+	clk_sys = (clock_control_subsys_t)(uintptr_t)cfg->clk_id;
+	err = clock_control_get_rate(cfg->clk_dev, clk_sys, &clock_rate);
+	if (err != 0) {
+		return err;
+	}
+
+	/* Approximate clock rate given ratios available */
+	div = spi_wch_get_br(clock_rate / config->frequency);
+#if CONFIG_SPI_LOG_LEVEL >= LOG_LEVEL_INF
+	uint32_t j = 2;
+	for (int i = 0; i < div; i++) {
+		j = j * 2;
+	}
+	LOG_INF("Selected divider %d, value %d, results in %d frequency", j, div, clock_rate / j);
+#endif
+	cfg->regs->CTLR1 |= (uint8_t)div << SPI_CTLR1_BR_POS;
+
+	data->ctx.config = config;
+
+	return 0;
+}
+
+static int spi_wch_transceive(const struct device *dev,
+			       const struct spi_config *config,
+			       const struct spi_buf_set *tx_bufs,
+			       const struct spi_buf_set *rx_bufs)
+{
+	const struct spi_wch_config *cfg = dev->config;
+	struct spi_wch_data *data = dev->data;
+	int err;
+	uint8_t rx;
+
+	spi_context_lock(&data->ctx, false, NULL, NULL, config);
+
+	err = spi_wch_configure(dev, config);
+	if (err != 0) {
+		goto done;
+	}
+
+	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
+
+	spi_context_cs_control(&data->ctx, true);
+
+	/* Start SPI *AFTER* setting CS */
+	cfg->regs->CTLR1 |= SPI_CTLR1_SPE;
+
+	while (spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx)) {
+		if (spi_context_tx_buf_on(&data->ctx)) {
+			cfg->regs->DATAR = *(uint8_t *)(data->ctx.tx_buf);
+			while(!(cfg->regs->STATR & SPI_STATR_TXE)) {}
```

Comment:
For consistency, use `while ((regs->STATR & SPI_START_TXE) != 0U) {
}`

Reply:
not quite... more like `while ((regs->STATR & SPI_START_TXE) == 0U) { }